### PR TITLE
Validate access list name length on creation

### PR DESCRIPTION
**Pull request by NickCellino**
Changelog: Added creation-time validation to disallow Access List names longer than 512 bytes.

Partially addresses gravitational/teleport-private#1867 (along with https://github.com/gravitational/teleport.e/pull/9227)

### Test Environment
- Local Teleport checkout on macOS

### Test Cases
- [x] `go test ./lib/accesslists -run 'TestAccessListValidateWithMembers_basic' -count=1`
- [x] Verify a new Access List creation with a name with 512 bytes is accepted.
- [x] Verify a new Access List creation with a name longer than 512 bytes is rejected.
    - Ran `tctl create <access-list.yaml>` specifying a name with 513 characters (which was rejected)
- [x] Verify an existing Access List with an oversized legacy name can still be updated.
    - Ran `tctl create <access-list.yaml>` specifying a name with 513 characters with `master` branch checked out (which was accepted)
    - Updated to latest version of this branch
    - Ran `tctl acl update <access-list-name> --description test-description` and it was accepted

### CODE REVIEW

**Comment by nixpig** [comment]
The manual test cases appear to test lengths of 511 bytes (accepted) and 513 bytes (rejected), but not the actual 512 bytes boundary. Is that intentional?

File: api/types/accesslist/accesslist.go [hunk 1]
```diff
@@ -78,6 +78,9 @@ func parseReviewFrequency(input string) ReviewFrequency {
 // MaxAllowedDepth is the maximum allowed depth for nested access lists.
 const MaxAllowedDepth = 10
 
+// MaxNameLength is the maximum permitted byte length of an access-list name.
```
Comment (nixpig) [inline api/types/accesslist/accesslist.go]:
```suggestion
// MaxNameLength is the maximum permitted byte length of an access list name.
```

File: lib/accesslists/validate_test.go [hunk 2]
```diff
@@ -756,6 +757,27 @@ func TestAccessListValidateWithMembers_basic(t *testing.T) {
...
 
+	t.Run("name length is validated on create", func(t *testing.T) {
+		t.Run("name length is too long", func(t *testing.T) {
+			accessList := newAccessList(t, strings.Repeat("x", accesslist.MaxNameLength+1), clock)
+			err := ValidateAccessListWithMembers(ctx, nil, accessList, nil, &mockAccessListAndMembersGetter{})
+			require.ErrorContains(t, err, "name is too long")
+		})
+
+		t.Run("name length is not too long", func(t *testing.T) {
+			accessList := newAccessList(t, strings.Repeat("x", accesslist.MaxNameLength-1), clock)
```
Comment (nixpig) [inline lib/accesslists/validate_test.go]: Could we update to just `accesslist.MaxNameLength` rather than `accesslist.MaxNameLength-1`, or add another test that tests the actual max name length?

File: lib/accesslists/validate.go [hunk 3]
```diff
@@ -58,6 +63,15 @@ func ValidateAccessListWithMembers(ctx context.Context, existingAccessList, acce
 	return nil
 }
 
+// validateAccessListCreate enforces validation rules that only apply during creation
```
Comment (nixpig) [inline lib/accesslists/validate.go]: nit:

```suggestion
// validateAccessListCreate enforces validation rules that only apply during creation.
```

File: lib/accesslists/validate_test.go [hunk 4]
```diff
@@ -756,6 +757,27 @@ func TestAccessListValidateWithMembers_basic(t *testing.T) {
 		})
 	})
 
+	t.Run("name length is validated on create", func(t *testing.T) {
```
Comment (nixpig) [inline lib/accesslists/validate_test.go:760]: nit: having this as a table would be more extensible if other validations are required in future, e.g. 

```go
[]struct{
  testName string
  accessListName string
  assertErr require.ErrorAssertionFunc
}{
  { "too long", strings.Repeat("x", accesslist.MaxNameLength+1), require.Error },
  { "just right", strings.Repeat("x", accesslist.MaxNameLength), require.NoError },
}
```

**Review by nixpig — CHANGES_REQUESTED** [review]
Couple of nits. 

Please can we add the test for an access list name _exactly_ `MaxNameLength`.

**Comment by NickCellino** [comment]
> Couple of nits.
> 
> Please can we add the test for an access list name _exactly_ `MaxNameLength`.

@nixpig done 👍

**Comment by NickCellino** [comment]
> The manual test cases appear to test lengths of 511 bytes (accepted) and 513 bytes (rejected), but not the actual 512 bytes boundary. Is that intentional?

Not intentional, I just didn't think to test at the actual boundary. Good catch! I've updated the manual test case as well as my unit tests to include this.

**Review by nixpig — APPROVED** [review]

File: lib/accesslists/validate_test.go [hunk 5]
```diff
@@ -756,6 +757,47 @@ func TestAccessListValidateWithMembers_basic(t *testing.T) {
...
+			},
+			{
+				name:           "short name",
+				accessListName: "x",
+			},
+		}
+		for _, test := range testCases {
+			t.Run(test.name, func(t *testing.T) {
+				accessList := newAccessList(t, test.accessListName, clock)
+				err := ValidateAccessListWithMembers(ctx, nil, accessList, nil, &mockAccessListAndMembersGetter{})
```
Comment (smallinsky) [inline lib/accesslists/validate_test.go:783]: Please add a test that the flow skips the validation if `existingAccessList` is not `nil`.

**Review by smallinsky — APPROVED** [review]

File: lib/accesslists/validate_test.go (hunk 5 above)
Comment (NickCellino) [reply lib/accesslists/validate_test.go:783]: That is already there, just below this subtest (`name length is not validated on update`)
